File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/integration/payment/adapter/PaymentResult.java
package it.giannibombelli.wsc2026.giftcard.application.integration.payment.adapter;

import it.giannibombelli.wsc2026.common.utils.Require;
import it.giannibombelli.wsc2026.giftcard.application.commands.ConfirmTopUp;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.payment.integration.PaymentResultIntegrationEvent;

import java.util.UUID;

/**
 * Anti-Corruption Layer che traduce la Published Language esposta da {@code payment}
 * nel command interno {@link ConfirmTopUp} del BC {@code giftcard}.
 */
public final class PaymentResult {

    public ConfirmTopUp adapt(PaymentResultIntegrationEvent event) {
        Require.requireArgument(event, "Payment result integration event");
        return switch (event) {
            case PaymentResultIntegrationEvent.PaymentAcceptedIntegrationEvent it ->
                new ConfirmTopUp(new GiftCardId(UUID.fromString(it.clientReference())), it.amount());
            case PaymentResultIntegrationEvent.PaymentRejectedIntegrationEvent ignored -> null;
            case PaymentResultIntegrationEvent.PaymentExpiredIntegrationEvent ignored -> null;
        };
    }
}

File: typescript/src/giftcard/application/integration/payment/adapter/paymentResult.ts
// Anti-Corruption Layer che traduce la Published Language esposta da `payment`
// nel command interno del BC `giftcard` per confermare una ricarica.
//
// Isola il modello di `giftcard` da eventuali evoluzioni del modello interno di `payment`:
// l'unico contratto condiviso è `PaymentResultIntegrationEvent`.

import type {
  PaymentAcceptedIntegrationEvent,
  PaymentResultIntegrationEvent,
} from '@/payment/integration/paymentResultIntegrationEvent.js';
import type { ConfirmTopUp } from '@/giftcard/application/commands/confirmTopUp.js';
import { confirmTopUp } from '@/giftcard/application/commands/confirmTopUp.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export class PaymentResult {
  adapt(event: PaymentResultIntegrationEvent): ConfirmTopUp | null {
    requireArgument(event, 'Payment result integration event');

    switch (event.kind) {
      case 'PaymentAcceptedIntegrationEvent':
        return this.adaptAccepted(event);
      case 'PaymentRejectedIntegrationEvent':
      case 'PaymentExpiredIntegrationEvent':
        return null;
    }
  }

  adaptAccepted(event: PaymentAcceptedIntegrationEvent): ConfirmTopUp {
    requireArgument(event, 'Payment accepted integration event');
    return confirmTopUp(
      new GiftCardId(Uuid.fromString(event.clientReference)),
      event.amount
    );
  }
}
